crates/pathfinder/src/sync/p2p/class_definitions.rs

```diff
@@ -0,0 +1,255 @@
+use std::num::NonZeroUsize;
+
+use anyhow::Context;
+use p2p::client::peer_agnostic::Class;
+use p2p::PeerData;
+use pathfinder_common::{BlockNumber, ClassHash};
+use pathfinder_storage::Storage;
+use starknet_gateway_types::class_definition::ClassDefinition;
+use starknet_gateway_types::class_hash::from_parts::{
+    compute_cairo_class_hash, compute_sierra_class_hash,
+};
+use tokio::task::spawn_blocking;
+
+#[derive(Debug, thiserror::Error)]
+pub(super) enum ClassDefinitionSyncError {
+    #[error(transparent)]
+    ClassDefinitionStreamError(#[from] anyhow::Error),
+    #[error("Invalid class definition layout")]
+    BadLayout(PeerData<(BlockNumber, ClassHash, serde_json::Error)>),
+    #[error("Class hash verification failed")]
+    BadClassHash(PeerData<(BlockNumber, ClassHash)>),
+}
+
+#[derive(Debug)]
+pub struct ClassWithLayout {
+    pub class: Class,
+    pub layout: ClassDefinition<'static>,
+}
+
+/// Returns the first block number which is missing at least one class definition, counting from genesis
+pub(super) async fn next_missing(
+    storage: Storage,
+    head: BlockNumber,
+) -> anyhow::Result<Option<BlockNumber>> {
+    spawn_blocking(move || {
+        let mut db = storage
+            .connection()
+            .context("Creating database connection")?;
+        let db = db.transaction().context("Creating database transaction")?;
+
+        if let Some(highest) = db
+            .highest_block_with_all_class_definitions_downloaded()
+            .context("Querying highest block with all class definitions")?
+        {
+            Ok((highest < head).then_some(highest + 1))
+        } else {
+            Ok(Some(BlockNumber::GENESIS))
+        }
+    })
+    .await
+    .context("Joining blocking task")?
+}
+
+pub(super) fn declared_class_counts_stream(
```

Comment:
Personally I would count the batch reading as a bit premature ito optimisation - a lot less code and testing surface if we could do something like

```rust
async fn get_count(storage: Storage, block: BlockNumber) -> Result<Count> {...}

...
    stream::iter(start..=stop)
        .then(|x| get_count(storage.clone(), x))
        ...
```

Reply:
No problem I can do it without batching too. But if so this also imo needs to be changed in state updates sync which rn does the batching too. Wdyt?

Reply:
I think its your call - if you're happy with the code then leave it :) 

I was just thinking if  we get to something like the above code then its really just no tests required - since its literally just a wrapped storage query which should already be tested.

Reply:
:thinking: After a second thought: a _oneshot_ `get_count` means cloning storage and then opening a new connection, all while occupying a thread from the tokio's blocking pool when the value fetched from the DB is literally an integer. 

So I'd stick to the current batching unless it's worse off performance wise (which I think we could check later on). As for testing once I get down to the tests, we'll _just_ use dependency injection.